indexdata.go

```diff
@@ -296,27 +300,57 @@ func (d *indexData) String() string {
 }
 
 // calculates an approximate size of indexData in memory in bytes.
-func (d *indexData) memoryUse() int {
-	sz := 0
+func (d *indexData) memoryUse() (ram, mmapped int) {
 	for _, a := range [][]uint32{
-		d.newlinesIndex, d.docSectionsIndex,
-		d.boundaries, d.fileNameIndex,
-		d.fileEndRunes, d.fileNameEndRunes,
-		d.fileEndSymbol, d.symbols.symKindIndex,
+		d.symbols.symKindIndex,
+		d.newlinesIndex,
+		d.docSectionsIndex,
+		d.boundaries,
+		d.fileEndRunes,
+		d.fileNameIndex,
+		d.fileEndSymbol,
+		d.fileNameEndRunes,
 		d.subRepos,
 	} {
-		sz += 4 * len(a)
+		ram += 4 * len(a)
 	}
-	sz += d.runeOffsets.sizeBytes()
-	sz += d.fileNameRuneOffsets.sizeBytes()
-	sz += len(d.languages)
-	sz += len(d.checksums)
-	sz += 2 * len(d.repos)
-	sz += 8 * len(d.runeDocSections)
-	sz += 8 * len(d.fileBranchMasks)
-	sz += d.ngrams.SizeBytes()
-	sz += 12 * len(d.fileNameNgrams) // these slices reference mmap-ed memory
-	return sz
+
+	for _, a := range [][]byte{
+		d.symbols.symContent,
+		d.symbols.symIndex,
+		d.symbols.symKindContent,
+		d.symbols.symMetaData,
+		d.runeDocSectionsRaw,
+		d.fileNameContent,
+		// fileNameNgrams
+		d.checksums,
+		d.languages,
+	} {
+		mmapped += len(a)
+	}
+
+	// This is the slowest part but the result is a non-negligible fraction of total.
+	// len(d.fileNameNgrams) is several thousand for an average simple shard.
+	for _, v := range d.fileNameNgrams {
+		mmapped += len(v)
+	}
```

Comment:
This is the most questionable part, I think.

Reply:
Can you elaborate? What exactly is questionable? The performance or whether it counts as mmaped?

Reply:
Perfromance.
Whether it counts is mmapped is not very transparent in the sense that we have to keep track of it by hand. Currently all []byte slices are mmapped, as far as I can tell.